src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java

```diff
@@ -329,6 +331,35 @@ private Object readResolve() {
             return getDisplayName();
         }
 
+        @Restricted(NoExternalUse.class) // for Jelly
+        public @CheckForNull String getEnclosingLabel() throws Exception {
+            FlowNode executorStepNode = context.get(FlowNode.class);
+            if (executorStepNode != null) {
+                for (StepExecution exec : executorStepNode.getExecution().getCurrentExecutions(true).get(1, TimeUnit.MINUTES)) {
+                    FlowNode runningNode = exec.getContext().get(FlowNode.class);
+                    if (runningNode != null) {
+                        // See if this step is inside our node {} block, and track the associated label.
+                        boolean match = false;
+                        String label = null;
+                        FlowNodeSerialWalker.EnhancedIterator it = new FlowNodeSerialWalker(runningNode).iterator();
```

Comment:
@jglick  Point of consideration: there may be many steps ahead of this before we see our first labeled or stage node.  Is there some form of implicit caching here that I am missing?  Otherwise, we need to make an effort not to walk the same flow graph with every UI view. 

My suggestion: store the ID & label of the last node in we got the enclosing label for.  If it is unchanged, return it.  Otherwise, only iterate through the walker until we hit changes in the node, or we hit the last node again.

Not perfect because of parallel blocks that can contain node blocks and labels, but better at least.